feat: Trino Authentications (#17593)
apache/superset: docs/src/pages/docs/Connecting to Databases/trino.mdx

## Trino

> SuperSet supports Trino >=352 via SQLAlchemy by using [sqlalchemy-trino](https://pypi.org/project/sqlalchemy-trino/) library.

Reviewer: ```suggestion
SuperSet supports Trino >=352 via SQLAlchemy by using the [sqlalchemy-trino](https://pypi.org/project/sqlalchemy-trino/) library.
```

> SuperSet supported Trino >=352 via SQLAlchemy by using [sqlalchemy-trino](https://pypi.org/project/sqlalchemy-trino/) library.

Reviewer 2: Nit: should this be:
```suggestion
Superset supports Trino >=352 via SQLAlchemy by using [sqlalchemy-trino](https://pypi.org/project/sqlalchemy-trino/) library.
```

### Connection String
The connection string format is as follows:
```
trino://{username}:{password}@{hostname}:{port}/{catalog}
```

> If you are running trino with docker on local machine, please use the following connection URL

Reviewer 2: nit:
```suggestion
If you are running Trino with docker on local machine, please use the following connection URL
```

```
trino://trino@host.docker.internal:8080
```

### Authentications
#### 1. Basic Authentication

> You can provide `username`/`password` in connection string or in `Secure Extra` field at `Advanced / Security`

Reviewer: ```suggestion
You can provide `username`/`password` in the connection string or in the `Secure Extra` field at `Advanced / Security`
```

* In Connection String
    ```
    trino://{username}:{password}@{hostname}:{port}/{catalog}
    ```

* In `Secure Extra` field
    ```json
    {
        "auth_method": "basic",
        "auth_params": {
            "username": "<username>",
            "password": "<password>"
        }
    }
    ```

> NOTE: if both are provided, `Secure Extra` always take higher priority.

Reviewer: ```suggestion
NOTE: if both are provided, `Secure Extra` always takes higher priority.
```

#### 2. Kerberos Authentication
In `Secure Extra` field, config as following example:
```json
{
    "auth_method": "kerberos",
    "auth_params": {
        "service_name": "SuperSet",
        "config": "/path/to/krb5.config",
        ...
    }
}
```

> All fields in `auth_params` are pass directly to [`KerberosAuthentication`](https://github.com/trinodb/trino-python-client/blob/0.306.0/trino/auth.py#L40) class.

Reviewer 2: nit:
```suggestion
All fields in `auth_params` are passed directly to [`KerberosAuthentication`](https://github.com/trinodb/trino-python-client/blob/0.306.0/trino/auth.py#L40) class.
```

> All fields in `auth_params` are passed directly to [`KerberosAuthentication`](https://github.com/trinodb/trino-python-client/blob/0.306.0/trino/auth.py#L40) class.

Reviewer: ```suggestion
All fields in `auth_params` are passed directly to the [`KerberosAuthentication`](https://github.com/trinodb/trino-python-client/blob/0.306.0/trino/auth.py#L40) class.
```

#### 3. JWT Authentication
Config `auth_method` and provide token in `Secure Extra` field
```json
{
    "auth_method": "jwt",
    "auth_params": {
        "token": "<your-jwt-token>"
    }
}
```

#### 4. Custom Authentication
```json
{
    "auth_method": "<your.module:AuthClass>",
    "auth_params": {
        ...
    }
}
```

> You can also provide custom authentication by provide reference to your `trino.auth.Authentication` class
> or create function (which return an `Authentication` instance).

Reviewer: ```suggestion
You can also provide custom authentication by providing reference to your `trino.auth.Authentication` class
or create a function (which returns an `Authentication` instance).
```

## Trino

> SuperSet supports Trino >=352 via SQLAlchemy by using the [sqlalchemy-trino](https://pypi.org/project/sqlalchemy-trino/) library.

Reviewer 2: The correct spelling is "Superset", not "SuperSet"
```suggestion
Superset supports Trino >=352 via SQLAlchemy by using the [sqlalchemy-trino](https://pypi.org/project/sqlalchemy-trino/) library.
```

Author: :) I just though `SuperSet` is right style

Author: Resolved

## Trino

Superset supports Trino >=352 via SQLAlchemy by using the [sqlalchemy-trino](https://pypi.org/project/sqlalchemy-trino/) library.

### Connection String
The connection string format is as follows:
```
trino://{username}:{password}@{hostname}:{port}/{catalog}
```

If you are running Trino with docker on local machine, please use the following connection URL
```
trino://trino@host.docker.internal:8080
```

### Authentications
#### 1. Basic Authentication
You can provide `username`/`password` in the connection string or in the `Secure Extra` field at `Advanced / Security`
* In Connection String
    ```
    trino://{username}:{password}@{hostname}:{port}/{catalog}
    ```

* In `Secure Extra` field
    ```json
    {
        "auth_method": "basic",
        "auth_params": {
            "username": "<username>",
            "password": "<password>"

>         }

Reviewer 3: @dungdm93 if people are currently passing username/password in their secure extra field, is this the same configuration, or will their existing configuration break?

Author: Basically, before this PR, passing username/password in the secure extra won't work, because trino wrap username/password in the `BasicAuthentication` object

Author: Basically, before this PR, passing username/password in the secure extra won't work, because trino wrap username/password in the `BasicAuthentication` object
